### PR TITLE
Fixes #27898 - Replace redux-form with formik

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/common/forms/TextField/TextField.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/TextField/TextField.js]: why not keep it as field?

File: webpack/assets/javascripts/react_app/components/common/forms/TextField/TextField.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/TextField/TextField.js]: i think you can write it as 
```js
 render={({ field, form }) => (
  <TextFieldInner....
  />
)
```

File: webpack/assets/javascripts/react_app/components/common/forms/validators.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/validators.js]: this should be identical to rails messages (for consistenticy across forms) see https://github.com/theforeman/foreman/blob/c599de355551b7756a683f202d894a9ebf7c7fa4/webpack/assets/javascripts/react_app/common/reduxFormI18n.js (which I guess should be removed?)

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: how about  using `try` - `catch` ? would make this more readable?

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: `reduce` looks a bit overkill here, maybe a helper function that creates errors object would be better ?

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: formik has its own `Form` and `ErrorMessage` components which reducing boilerplate, does it make sense to consume these instead of our own implementation ?

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: I like the `validationSchema` which makes validation cleaner

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @xprazak2 !  looks awesome and  Formik looks a nice alternative to redux-forms!

I think it would be nice to have a generic wrapper, so a consumer doesn't need to work directly with formik (plus adding a form component generator ? )   Also I like the `SchemaValidation` and `ErrorMessage` built-in components which can make the code even cleaner.

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: Not sure about readability but submitForm is asynchronous.

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: I use `Form` from formik now, but it does not provide significant benefits. `ErrorMessage` provides just a basic wiring, we already have a highly customized way to display errors.

**Comment by xprazak2** [comment]
New additions:

* validations using schema, it even helped solving the problem with errors on initialization. 
* added a `ForemanForm` component that makes it more straightforward to create a form

**Comment by xprazak2** [comment]
I updated tests, improved style.

**Comment by xprazak2** [comment]
Rebased and opened PR for foreman-js updating the depndencies.

File: webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js:66]: why remove the tests?

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js]: why is this required?

File: webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap:15]: is this a bug?

File: webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap:92]: is this a bug?

File: webpack/assets/javascripts/react_app/components/common/forms/validators.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/forms/validators.js]: would this combine the api errors as well (e.g. name is already been taken?)

**Review by ohadlevy — COMMENTED** [review]
thanks for doing this! asked more questions as I wasn't able to get the env running (needed to meddle with foreman-js to get it working)

File: webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js:66]: `spy` was refusing to recognize it had been called, But I figured out how to make tests pass and put them back - I needed to [wait](https://github.com/jaredpalmer/formik/blob/master/test/Formik.test.tsx#L209) with [react-testing-library](https://github.com/testing-library/react-testing-library#wait) because the test was exiting too early.

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js]: To inject additional props to children - so that we can disable fields when submitting a form.

File: webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap:15]: It keeps the 'Cancel' button enabled even if user haven't touched the form.

File: webpack/assets/javascripts/react_app/components/common/forms/validators.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/validators.js]: The original thought was to combine error messages from validators that were applied to a particular form field. It is no longer needed because we use schema validation instead, removing.

**Comment by xprazak2** [comment]
@ohadlevy, thank you for your comments. `npm install` should be now enough to get it working, I updated package.json with versions of foreman-js that contain the new dependencies (thank you @sharvit for building it!).

Tests will not pass as @testing-library/react is not included there yet.

File: webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/__snapshots__/Actions.test.js.snap:92]: The diff looks misleading because snapshots moved a bit as result of changes in tests, but the test snapshot does have a spinner when inspecting the snapshot file.

File: webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js]: You might be able to use the `IntegrationTestHelper.flushAllPromises();` instead:
https://sharvit.github.io/react-redux-test-utils/class/src/IntegrationTestHelper.js~IntegrationTestHelper.html#static-method-flushAllPromises

**Comment by sharvit** [comment]
> Tests will not pass as @testing-library/react is not included there yet.

Shouldn't it suppose to be a dev dependency anyway?

**Comment by xprazak2** [comment]
[test foreman]
[test katello]

**Comment by jeremylenz** [comment]
@xprazak2 @sharvit I was able to get foreman running by changing `package.json` dependencies from 
```
"@theforeman/vendor": "^1.0.0-formik.0",
```
to
```
"@theforeman/vendor": "1.0.0-formik.0",
```
but Storybook is still not working.. it's saying 
```
Module build failed: 
@import '~@theforeman/vendor/scss/vendor';
^
      File to import not found or unreadable: ~@theforeman/vendor/scss/vendor.
      in /home/vagrant/foreman/webpack/stories/index.scss (line 6, column 1)
ERROR in ./webpack/assets/javascripts/react_app/components/common/forms/Form.stories.js
Module not found: Error: Can't resolve 'formik' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/common/forms'
ERROR in ./webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js
Module not found: Error: Can't resolve 'formik' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/common/forms/ForemanForm'
ERROR in ./webpack/assets/javascripts/react_app/components/common/forms/RadioButtonGroup/RadioButtonGroup.js
Module not found: Error: Can't resolve 'formik' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/common/forms/RadioButtonGroup'
ERROR in ./webpack/assets/javascripts/react_app/components/common/forms/TextField/TextField.js
Module not found: Error: Can't resolve 'formik' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/common/forms/TextField'
ERROR in ./webpack/assets/javascripts/react_app/components/bookmarks/form/BookmarkForm.js
Module not found: Error: Can't resolve 'yup' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/bookmarks/form'
ERROR in ./webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js
Module not found: Error: Can't resolve 'yup' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/common/forms/ForemanForm'
```

**Comment by sharvit** [comment]
@jeremylenz it is because the version `1.0.0-formik.0` doesn't contain the storybook fix.

@xprazak2 You can rebase your vendor PR so I can release `1.0.0-formik.1` with the storybook fix.

Or maybe better, If we feel comfortable with `formik` and this is the direction we want to go, I would be ok with releasing a new official vendor with `redux-form` together with `formik` so it will be easier to push it forward with the development and the review process.

**Comment by jeremylenz** [comment]
@sharvit I'm still unable to run storybook:
```
ERROR in ./webpack/assets/javascripts/react_app/components/common/forms/Form.stories.js
Module not found: Error: Can't resolve 'formik' in '/home/vagrant/foreman/webpack/assets/javascripts/react_app/components/common/forms'
```
What do I need to test this?

**Comment by sharvit** [comment]
`@theforeman/vendor` version `1.5.0` was released with `formik`, I assume it should fix the storybook:
https://github.com/theforeman/foreman-js/pull/36#issuecomment-531671040

**Comment by xprazak2** [comment]
Rebased and updated the `@theforeman/vendor` version to 1.5.0, the dependencies now resolve correctly.

**Comment by jeremylenz** [comment]
@xprazak2 Storybook is working now, thanks!

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/index.js]: Someone told me once that in Foreman the convention is to have this file named `ForemanForm.js` and then have another file `index.js` that imports and exports it.

File: webpack/assets/javascripts/react_app/components/Bookmarks/BookmarksActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/Bookmarks/BookmarksActions.js:6]: Thanks for making our codebase better with descriptive filenames :)

File: webpack/assets/javascripts/react_app/components/Bookmarks/BookmarksConstants.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/Bookmarks/BookmarksConstants.js]: Why only this one singular and the rest plural?  Would it be better to call it `BOOKMARKS_FORM_SUBMITTED`?

File: webpack/assets/javascripts/react_app/components/Bookmarks/__tests__/BookmarksReducer.test.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/Bookmarks/__tests__/BookmarksReducer.test.js]:
```suggestion
  'should update after form submission': {
```

File: webpack/assets/javascripts/react_app/components/Bookmarks/__tests__/BookmarksSelectors.test.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/Bookmarks/__tests__/BookmarksSelectors.test.js:16]: This is cool :) I like how you use the selectors here

File: webpack/assets/javascripts/react_app/components/Bookmarks/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/Bookmarks/index.js:1]: Same comment as before about the filename :)

File: webpack/assets/javascripts/react_app/components/common/forms/Actions.test.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/common/forms/Actions.test.js]: Maybe change to `'should show disabled buttons'` ? I think it's helpful if the English grammar is correct when using `it`.

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js:19]: I was surprised that my live form values were showing up here in "props," since form state is usually kept in React component "state" and not "props."  But "props" here refers to the props passed down by Formik to your form, I guess?

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js:61]: No, I want a hamburger thx :laughing:

**Review by jeremylenz — DISMISSED** [review]
Looks awesome!  Nitpicks etc. below.  Looking forward to using this :)

File: webpack/assets/javascripts/react_app/components/Bookmarks/BookmarksConstants.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/Bookmarks/BookmarksConstants.js]: Mainly because it is a form for one bookmark and also what [submitForm](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/redux/actions/common/forms.js#L45) does with `item`, but I made changes to have it in plural.

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js:19]: Yes, these are passed down by Formik.

File: webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.stories.js:61]: :hamburger: :wink:

File: webpack/assets/javascripts/react_app/components/Bookmarks/index.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/Bookmarks/index.js:1]: I think this is ok to keep as is - `index.js` is actually meant to export the connected component if the component needs to access the store. That way we have `MyComponent/MyComponent.js` without wiring to store (useful for tests) and we connect it to redux in `MyComponent/index.js` - [SearchBar](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/SearchBar/index.js) or [Breadcrumbs](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/BreadcrumbBar/index.js) have this structure.

**Comment by xprazak2** [comment]
@jeremylenz, thank you for review, I added changes based on your comments. I kept it as a separate commit to make it obvious what changed, but it will be good to squash before merge

I also attached a redmine issue.

File: webpack/assets/javascripts/react_app/components/Bookmarks/index.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/Bookmarks/index.js:1]: @xprazak2 Right, I think I might have been trying to say the same thing :)

**Review by jeremylenz — DISMISSED** [review]
:+1: :+1:

**Comment by sharvit** [comment]
@xprazak2 notice the linting error:
https://travis-ci.org/theforeman/foreman/jobs/589406073#L2284-L2285

File: webpack/assets/javascripts/react_app/components/bookmarks/form/index.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/bookmarks/form/index.js]: @xprazak2 Anything that returns a promise can be used either with `then` / `catch` or with `async/await` in a `try` - `catch`; should make no difference :)

**Review by jeremylenz — DISMISSED** [review]
This looks good to me after test coverage etc. checks are addressed.

**Review by amirfefer — COMMENTED** [review]
Tested with bookmarks and it works as expected :+1: 

Could you add a "Creating a form" markdown documentation? 
please notice that tests are failing

**Comment by xprazak2** [comment]
I added docs and rebased on top of API middleware.

**Review by amirfefer — DISMISSED** [review]
I'm in favor to merge it (after fixing the snapshot), works well, well tested and documented (plus we just started a new branch :) )
Awesome work @xprazak2! Thanks :+1:

File: webpack/stories/docs/creatingForm.md
Comment (amirfefer) [inline webpack/stories/docs/creatingForm.md:29]: Hopefully it's a pistachio flavor :ice_cream:  :)

File: webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap]: Oh snap...  this snap looks broken

**Comment by jeremylenz** [comment]
@xprazak2 what is the status of this one?  I think this will have merge conflicts with the work in progress on #28161, but want to get this merged so I can deal with it

File: webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap]: This snap looks strange, but I think it is actually correct, the component returns a Formik field...

**Comment by xprazak2** [comment]
I rebased on the latest develop.

File: webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap]: but it returns the same formik component on different given props, it looks like a redundant snap to me, if formik handles it by an inner implementation, IMO we don't need to test it in foreman

**Comment by xprazak2** [comment]
All green, finally.

**Comment by xprazak2** [comment]
Rebased, could I get ack from packaging?

**Comment by xprazak2** [comment]
@ekohl, could you ACK please? The only change in package.json is for tests.

File: webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/forms/TextField/__snapshots__/TextField.test.js.snap]: ?

**Review by ekohl — DISMISSED** [review]
:+1: from packaging since `react-redux-test-utils` is skipped as a dependency. Don't know about the actual JS change so I'll leave the merge to someone with UI skills.

**Review by amirfefer — COMMENTED** [review]
could you please remove [this snap](https://github.com/theforeman/foreman/pull/6832/files#diff-1e682f7ba6b24e8cda552653ca759fe2R1) ?

**Comment by xprazak2** [comment]
I removed the snap.

**Review by mmoll — APPROVED** [review]

**Review by amirfefer — APPROVED** [review]
Thanks @xprazak2 ! :+1: 
code wise and works well (bookmarks and stories)

**Pull request merged**
